cpp/src/strings/convert/convert_floats.cu

```diff
@@ -130,10 +130,17 @@ __device__ inline double stod(string_view const& d_str)
   else if (exp_ten < std::numeric_limits<double>::min_exponent10)
     return double{0};
 
+  double base = sign * static_cast<double>(digits);
+
   exp_ten += 1 - num_digits;
-  // exp10() is faster than pow(10.0,exp_ten)
+  // extra floating-point division needed only in extreme range (e-287 - e-307)
```

Comment:
Are these subnormal values? I'd like to use a precise floating point term to describe these.

Reply:
No, these are not subnormal values. Are you have an issue with wording in the comment?

Reply:
I did a deep dive and this problem is arising due to subnormal values.

Here's an explanation of the problem and one way we could fix it. I'm going to drop a bunch of links here just so I have a place to track my work, but the only truly important link is the last one (solution, just the algorithm).

Examples of subnormal values: https://godbolt.org/z/srbxxvhdf

The snippet below is returning 0 for normal values between 1e-307 and `DBL_MIN ~= 2.22507e-308`, as well as subnormal values between `DBL_MIN` and `DBL_TRUE_MIN ~= 4.94066e-324`. Ideally, these values should not parse as 0.

```cpp
  } else if (exp_ten < std::numeric_limits<double>::min_exponent10) {
    return double{0};
  }
```

Something close to the previous algorithm (minor adjustments for stdlib compat, no nan/inf support): https://godbolt.org/z/bvbvhY6Wq
- I copied the test case "1.7976931348623157e-307", which parses as 0 ❌

Something close to the current algorithm (minor adjustments for stdlib compat, no nan/inf support): https://godbolt.org/z/cocWMq1Mh
- Test case "1.7976931348623157e-307" parses correctly ✔️ 
- Test case "4e-308" is normal, parses as 0 ❌ 
- Test case "1e-308" is subnormal, parses as 0 ❌ 

Something that handles subnormal values:
(with comments / print statements): https://godbolt.org/z/ds9ssW9aK
(just the algorithm): https://godbolt.org/z/hGqKqKqo4
- Test case "1.7976931348623157e-307" parses correctly ✔️ 
- Test case "4e-308" is normal, parses correctly ✔️ 
- Test case "1e-308" is subnormal, parses correctly ✔️ 
- Test case "3.3333333333e-320" is subnormal, loses precision around 4th decimal place as expected. ✔️
- Test case "4.940656458412465441765688e-324" is a representation of the smallest subnormal, parses as expected. ✔️ 
- Test case "1.e-324" is less than the smallest subnormal, and returns 0 as expected. ✔️

Suggested code (I will also make a GitHub suggestion with this snippet):
```cpp
  if (exp_ten > std::numeric_limits<double>::max_exponent10)
    return sign > 0 ? std::numeric_limits<double>::infinity()
                    : -std::numeric_limits<double>::infinity();

  double base = sign * static_cast<double>(digits);

  exp_ten += 1 - num_digits;
  // If 10^exp_ten would result in a subnormal value, the base and
  // exponent should be adjusted so that 10^exp_ten is a normal value
  int const subnormal_shift = std::numeric_limits<double>::min_exponent10 - exp_ten;
  if (subnormal_shift > 0) {
    // Handle subnormal values. Ensure that both base and exponent are
    // normal values before computing their product.
    base = base / exp10(static_cast<double>(num_digits - 1 + subnormal_shift));
    exp_ten += num_digits - 1;  // adjust exponent
    double const exponent = exp10(static_cast<double>(exp_ten + subnormal_shift));
    return base * exponent;
  }

  double const exponent = exp10(static_cast<double>(std::abs(exp_ten)));
  return exp_ten < 0 ? base / exponent : base * exponent;
}
```

I can't promise that this will have any nice small bound on ULPs (precision loss) but my brief testing indicates this is probably close enough and fixes the core of the problem. Importantly, I tried to make it so that this change will not harm the precision for parsing normal values.

Reply:
This is awesome. Thanks for spending so much time on this. Your analysis and implementation looks good to me. This has one less if-statement, the same number of floating-point operations and now handles subnormal numbers! I will include this code and test cases into this PR.

Reply:
Woohoo! 🚀